### PR TITLE
make the parser algorithm more readable by adding helper methods

### PR DESCRIPTION
**Pull request by dwightguth**
@radumereuta please review.

### CODE REVIEW

File: src/javasources/KTool/src/org/kframework/parser/concrete2/Parser.java [hunk 1]
```diff
@@ -229,14 +233,34 @@ public int hashCode() {
...
+                this.ntCall().exitStateReturns.add(this);
             }
         }
 
         public int hashCode() {
             return this.key.hashCode();
         }
+        
+        public State state() {
+            return this.key.stateCall.key.state;
```
Comment (adamsmd) [inline src/javasources/KTool/src/org/kframework/parser/concrete2/Parser.java:246]: Why is there a "StateCall.stateBegin()" but not a "StateCall.state()"?  (This question also goes for the other fields of NonTerminalCall, StateCall and StateReturn.)

File: src/javasources/KTool/src/org/kframework/parser/concrete2/Parser.java [hunk 2]
```diff
@@ -390,6 +414,24 @@ public ParseState(CharSequence input) {
...
             columns[input.length()] = c;
         }
+        
+        void enqueueStateReturn(StateCall stateCall, int stateEnd)  {
+            this.stateReturnWorkList.enqueue(
+                    this.stateReturns.get(
+                            new StateReturn.Key(stateCall, stateEnd)));
+        }
+        
+        NonTerminalCall getNtCall(NonTerminal nt, int ntBegin) {
```
Comment (adamsmd) [inline src/javasources/KTool/src/org/kframework/parser/concrete2/Parser.java:424]: Shouldn't there be a getStateReturn() to correspond to getNtCall() and getStateCall()?

File: src/javasources/KTool/src/org/kframework/parser/concrete2/Parser.java [hunk 3]
```diff
@@ -744,34 +783,30 @@ private void workListStep(StateReturn stateReturn) {
...
                 s.stateReturns.get(new StateReturn.Key(
-                    s.stateCalls.get(new StateCall.Key(
-                        s.ntCalls.get(new NonTerminalCall.Key(
-                            ((Grammar.NonTerminalState) stateReturn.key.stateCall.key.state).child,
-                            stateReturn.key.stateCall.key.stateBegin)),
-                        stateReturn.key.stateEnd,
-                        ((Grammar.NonTerminalState) stateReturn.key.stateCall.key.state).child.exitState)),
-                    stateReturn.key.stateEnd)).function);
+                        s.getStateCall(ntState.child, stateReturn.stateBegin(),
+                                stateReturn.stateEnd(), ntState.child.exitState),
```
Comment (adamsmd) [inline src/javasources/KTool/src/org/kframework/parser/concrete2/Parser.java:808]: This is misleadingly indented.  Line 808 is part of the getStateCall(), but line 809 is outside the getStateCall().  I think line 808 should be indented to align with the start of "ntState.child" and at least have line 809 be indented to align with the start of "s.getStateCall".

File: src/javasources/KTool/src/org/kframework/parser/concrete2/Parser.java [hunk 4]
```diff
@@ -390,6 +414,24 @@ public ParseState(CharSequence input) {
...
+            this.stateReturnWorkList.enqueue(
+                    this.stateReturns.get(
+                            new StateReturn.Key(stateCall, stateEnd)));
+        }
+        
+        NonTerminalCall getNtCall(NonTerminal nt, int ntBegin) {
+            return ntCalls.get(new NonTerminalCall.Key(nt, ntBegin));
+        }
+        
+        StateCall getStateCall(NonTerminalCall ntCall, int stateBegin, State state) {
```
Comment (adamsmd) [inline src/javasources/KTool/src/org/kframework/parser/concrete2/Parser.java:428]: If you are using solely these higher level functions for accessing ntCalls, stateCalls and stateReturns, you might make ntCalls, stateCalls and stateReturns private.

**Comment by radumereuta** [comment]
I would like to do this for myself at some point.
I will close it for now.

**Pull request closed without merging**